File: js/commands/ops.js
/* ROOT OF CRIME – process / disk / net / cron */

Object.assign(Terminal, {
  _cmdPs(args) {
    const procs = this._liveProcesses();
    if (!procs.length) return this._ok('  PID TTY          TIME CMD\n    1 ?            00:00:12 init');
    const long = args.some((a) => /a|u|x|e|f/.test(a.replace(/^-/, ''))) || args.includes('aux') || args.includes('-aux') || args.includes('-ef');
    if (long) {
      const lines = ['USER         PID %CPU %MEM    TTY      STAT START   TIME COMMAND'];
      const htmlLines = [this._escape(lines[0])];
      procs.forEach((p) => {
        const line = `${p.user.padEnd(12)}${String(p.pid).padEnd(6)}${String(p.cpu).padEnd(5)}${String(p.mem).padEnd(5)}  ${String(p.tty).padEnd(7)} S    ${String(p.start).padEnd(7)}${p.time} ${p.cmd}`;
        lines.push(line);
        if (p.highlight) htmlLines.push(`<span class="highlight">${this._escape(line)}</span>`);
        else htmlLines.push(this._escape(line));
      });
      return { stdout: lines.join('\n') + '\n', stderr: '', code: 0, html: htmlLines.join('\n') + '\n' };
    }
    const lines = ['  PID TTY          TIME CMD'];
    procs.forEach((p) => {
      lines.push(` ${String(p.pid).padStart(5)} ${String(p.tty).padEnd(12)} ${p.time} ${p.cmd.split(' ')[0]}`);
    });
    return this._ok(lines.join('\n'));
  },

  _cmdKill(args) {
    if (!this.ctx?.processes) return this._err('kill: no process table in this context');
    let signal = 'TERM';
    const rest = [];
    args.forEach((a) => {
      if (['-9', '-KILL', '-SIGKILL'].includes(a)) signal = 'KILL';
      else if (['-15', '-TERM', '-SIGTERM'].includes(a)) signal = 'TERM';
      else rest.push(a);
    });
    if (!rest.length) return this._err('kill: usage: kill [-9] <PID>');
    const pid = parseInt(rest[0], 10);
    if (!pid) return this._err('kill: usage: kill [-9] <PID>');
    return this._killPid(pid, signal);
  },

  _cmdPkill(args) {
    if (!this.ctx?.processes) return this._err('pkill: no process table in this context');
    const pattern = (args[0] || '').toLowerCase();
    if (!pattern) return this._err('pkill: usage: pkill <pattern>');
    const hits = this._liveProcesses().filter((p) => p.cmd.toLowerCase().includes(pattern) || p.cmd.split('/').pop().toLowerCase().includes(pattern));
    if (!hits.length) return this._err(`pkill: no process found matching '${args[0]}'`);
    const results = [];
    hits.forEach((p) => {
      const r = this._killPid(p.pid, 'TERM');
      if (r.stdout) results.push(r.stdout.trim());
    });
    return this._ok(results.join('\n'));
  },

  _cmdDf(args) {
    const guest = this.attachedGuest && this.attachedGuest();
    if (guest) return this._ok(Virt.dfTable(guest));
    const used = VFS.sizeOf(this.vfs);
    const total = this.ctx?.diskTotal || 512000;
    const avail = Math.max(0, total - used);
    const pct = Math.min(99, Math.round((used / total) * 100));
    const human = (args || []).some((a) => a === '-h' || a === '-H');
    const fmt = (n) => {
      if (!human) return `${Math.max(0, Math.round(n / 1024))}K`;
      if (n < 1024) return `${n}B`;
      if (n < 1024 * 1024) return `${Math.round(n / 1024)}K`;
      return `${(n / (1024 * 1024)).toFixed(1)}M`;
    };
    if (human) {
      return this._ok(
        'Filesystem      Size  Used Avail Use% Mounted on\n' +
        `/dev/sda1      ${fmt(total).padStart(5)} ${fmt(used).padStart(5)} ${fmt(avail).padStart(5)}  ${String(pct).padStart(3)}% /`
      );
    }
    return this._ok(
      'Filesystem     1K-blocks    Used Available Use% Mounted on\n' +
      `/dev/sda1      ${fmt(total).padStart(9)} ${fmt(used).padStart(7)} ${fmt(avail).padStart(9)}  ${String(pct).padStart(3)}% /`
    );
  },

  _cmdDu(args) {
    const human = args.includes('-h') || args.includes('-sh') || args.includes('-hs');
    const summarize = args.includes('-s') || args.includes('-sh') || args.includes('-hs');
    const paths = args.filter((a) => !a.startsWith('-'));
    const targets = paths.length ? paths : ['.'];
    const format = (n) => {
      if (!human) return String(Math.max(1, Math.round(n / 1024)));
      if (n < 1024) return `${n}B`;
      if (n < 1024 * 1024) return `${Math.round(n / 1024)}K`;
      return `${(n / (1024 * 1024)).toFixed(1)}M`;
    };
    const lines = [];
    for (const target of targets) {
      const res = VFS.resolve(this.vfs, this.cwd, target, this.home);
      if (!res) return this._err(`du: cannot access '${target}': No such file or directory`);
      if (summarize || res.node.type === 'file' || targets.length > 1) {
        lines.push(`${format(VFS.sizeOf(res.node))}\t${target}`);
        continue;
      }
      Object.entries(res.node.children || {}).forEach(([name, child]) => {
        lines.push(`${format(VFS.sizeOf(child))}\t${name}`);
      });
      lines.push(`${format(VFS.sizeOf(res.node))}\t.`);
    }
    return this._ok(lines.join('\n'));
  },

  _cmdNetstat() {
    const conns = this.ctx?.connections || [
      { proto: 'tcp', local: '0.0.0.0:22', remote: '0.0.0.0:*', state: 'LISTEN', proc: 'sshd' },
      { proto: 'tcp', local: '127.0.0.1:631', remote: '0.0.0.0:*', state: 'LISTEN', proc: 'cupsd' }
    ];
    const lines = ['Proto Local Address           Foreign Address         State       PID/Program name'];
    conns.forEach((c) => {
      const line = `${c.proto.padEnd(5)} ${c.local.padEnd(22)} ${c.remote.padEnd(23)} ${c.state.padEnd(11)} ${c.proc}`;
      lines.push(c.highlight ? line : line);
    });
    if (this.ctx) this.ctx.usedNetstat = true;
    const html = lines.map((line, i) => {
      if (i > 0 && conns[i - 1]?.highlight) return `<span class="highlight">${this._escape(line)}</span>`;
      return this._escape(line);
    }).join('\n') + '\n';
    return { stdout: lines.join('\n') + '\n', stderr: '', code: 0, html };
  },

  _cmdLast() {
    const extra = this.ctx?.lastLog || '';
    const base =
      'root     tty1         Fri Aug 14 19:02   still logged in\n' +
      'chief    pts/0        Fri Aug 14 18:55   still logged in\n' +
      'miller   pts/2        Fri Aug 14 17:10 - 17:44  (00:34)\n';
    if (this.ctx) this.ctx.usedLast = true;
    return this._ok(base + extra);
  },

  _cmdCrontab(args) {
    if (!args.includes('-l') && args[0] !== '-l') {
      return this._err('crontab: usage: crontab -l');
    }
    const file = VFS.resolve(this.vfs, '/', '/etc/crontab');
    const extraDir = VFS.resolve(this.vfs, '/', '/etc/cron.d');
    let out = file ? file.node.content : '';
    if (extraDir && extraDir.node.type === 'dir') {
      Object.entries(extraDir.node.children).forEach(([name, node]) => {
        out += `\n# /etc/cron.d/${name}\n${node.content}`;
      });
    }
    if (this.ctx) this.ctx.usedCrontab = true;
    return this._ok(out);
  },

  _cmdPasswd(args) {
    const target = args[0] || 'root';
    if (this.ctx?.allowPasswdFor && (target === this.ctx.allowPasswdFor || args.length === 0 && this.ctx.allowPasswdFor === 'root')) {
      this.ctx.passwordChanged = this.ctx.passwordChanged || [];
      this.ctx.passwordChanged.push(this.ctx.allowPasswdFor);
      return this._ok(`Changing password for ${this.ctx.allowPasswdFor}.\nNew password: ********\npasswd: password updated successfully`);
    }
    if (target === 'coffee' || target === 'beantek') {
      this.ctx.passwordChanged = this.ctx.passwordChanged || [];
      this.ctx.passwordChanged.push('coffee');
      return this._ok('Changing password for coffee.\nNew password: ********\npasswd: password updated successfully');
    }
    return this._err('passwd: Authentication token manipulation error\n(This console only lets you rotate the appliance account.)');
  },

  _cmdPing(args) {
    const raw = (args[0] || '').trim();
    if (!raw || raw.startsWith('-')) return this._err('usage: ping HOST');
    if (typeof Infra === 'undefined') return this._err('ping: network is down');
    Infra.boot();
    const id = Infra.resolve(raw);
    const rec = Infra.catalog(id);
    if (!rec) return this._err('ping: ' + raw + ': Name or service not known');
    return this._ok(
      'PING ' + rec.id + ' (' + rec.addr + ') 56(84) bytes of data.\n' +
      '64 bytes from ' + rec.id + ' (' + rec.addr + '): icmp_seq=1 ttl=64 time=0.4 ms\n' +
      '64 bytes from ' + rec.id + ' (' + rec.addr + '): icmp_seq=2 ttl=64 time=0.3 ms\n' +
      '--- ' + rec.id + ' ping statistics ---\n' +
      '2 packets transmitted, 2 received, 0% packet loss'
    );
  }
});

Terminal.define('ps', function (args) { return this._cmdPs(args); });
Terminal.define('kill', function (args) { return this._cmdKill(args); });
Terminal.define('pkill', function (args) { return this._cmdPkill(args); });
Terminal.define('df', function (args) { return this._cmdDf(args); });
Terminal.define(['netstat', 'ss'], function (args) { return this._cmdNetstat(args); });
Terminal.define('du', function (args) { return this._cmdDu(args); });
Terminal.define('last', function () { return this._cmdLast(); });
Terminal.define('who', function () {
  return this._ok('root     tty1         2026-08-14 19:02\nchief    pts/0        2026-08-14 18:55');
});
Terminal.define('crontab', function (args) { return this._cmdCrontab(args); });
Terminal.define('passwd', function (args) { return this._cmdPasswd(args); });
Terminal.define('ping', function (args) { return this._cmdPing(args); });

File: crates/roc-kernel/src/shell/commands/ops.rs
//! Process / disk / net / cron. Matches `js/commands/ops.js`.

use crate::lan;
use crate::proc::{default_connections, Proc};
use crate::vfs::{self, size_of};

use super::super::exec::Shell;
use super::super::output::Output;

impl Shell {
    fn live_processes(&self) -> Vec<&Proc> {
        self.ctx
            .processes
            .as_ref()
            .map(|ps| ps.iter().filter(|p| !p.dead).collect())
            .unwrap_or_default()
    }

    fn kill_pid(&mut self, pid: i32, signal: &str) -> Output {
        if self.ctx.processes.is_none() {
            return Output::err("kill: no process table in this context");
        }
        let snapshot = {
            let procs = self.ctx.processes.as_mut().unwrap();
            let Some(proc) = procs.iter_mut().find(|p| p.pid == pid && !p.dead) else {
                return Output::err(format!("kill: ({pid}) - No such process"));
            };
            if pid == 1 || proc.protected {
                return Output::err(format!("kill: ({pid}) - Operation not permitted"));
            }
            proc.dead = true;
            proc.clone()
        };
        self.ctx.killed.push(snapshot.clone());
        let name = snapshot
            .cmd
            .split(' ')
            .next()
            .unwrap_or("")
            .rsplit('/')
            .next()
            .unwrap_or("");
        let word = if signal == "KILL" { "Killed" } else { "Terminated" };
        Output::ok(format!("[1]  + {word}                 {name}"))
    }

    pub(crate) fn cmd_ps(&mut self, args: &[String]) -> Output {
        let procs = self.live_processes();
        if procs.is_empty() {
            return Output::ok("  PID TTY          TIME CMD\n    1 ?            00:00:12 init");
        }
        let long = args.iter().any(|a| {
            let flags = a.trim_start_matches('-');
            flags.chars().any(|c| matches!(c, 'a' | 'u' | 'x' | 'e' | 'f'))
                || a == "aux"
                || a == "-aux"
                || a == "-ef"
        });
        if long {
            let mut lines = vec!["USER         PID %CPU %MEM    TTY      STAT START   TIME COMMAND".to_string()];
            for p in &procs {
                let line = format!(
                    "{:<12}{:<6}{:<5}{:<5}  {:<7} S    {:<7}{} {}",
                    p.user, p.pid, p.cpu, p.mem, p.tty, p.start, p.time, p.cmd
                );
                lines.push(line);
            }
            return Output::ok(lines.join("\n"));
        }
        let mut lines = vec!["  PID TTY          TIME CMD".to_string()];
        for p in &procs {
            let cmd0 = p.cmd.split(' ').next().unwrap_or("");
            lines.push(format!(
                " {:>5} {:<12} {} {cmd0}",
                p.pid, p.tty, p.time
            ));
        }
        Output::ok(lines.join("\n"))
    }

    pub(crate) fn cmd_kill(&mut self, args: &[String]) -> Output {
        if self.ctx.processes.is_none() {
            return Output::err("kill: no process table in this context");
        }
        let mut signal = "TERM";
        let mut rest = Vec::new();
        for a in args {
            if matches!(a.as_str(), "-9" | "-KILL" | "-SIGKILL") {
                signal = "KILL";
            } else if matches!(a.as_str(), "-15" | "-TERM" | "-SIGTERM") {
                signal = "TERM";
            } else {
                rest.push(a.as_str());
            }
        }
        if rest.is_empty() {
            return Output::err("kill: usage: kill [-9] <PID>");
        }
        let pid: i32 = match rest[0].parse() {
            Ok(n) if n != 0 => n,
            _ => return Output::err("kill: usage: kill [-9] <PID>"),
        };
        self.kill_pid(pid, signal)
    }

    pub(crate) fn cmd_pkill(&mut self, args: &[String]) -> Output {
        if self.ctx.processes.is_none() {
            return Output::err("pkill: no process table in this context");
        }
        let pattern = args.first().map(|s| s.to_ascii_lowercase()).unwrap_or_default();
        if pattern.is_empty() {
            return Output::err("pkill: usage: pkill <pattern>");
        }
        let pids: Vec<i32> = self
            .live_processes()
            .into_iter()
            .filter(|p| {
                let cmd = p.cmd.to_ascii_lowercase();
                let base = cmd.rsplit('/').next().unwrap_or("");
                cmd.contains(&pattern) || base.contains(&pattern)
            })
            .map(|p| p.pid)
            .collect();
        if pids.is_empty() {
            return Output::err(format!(
                "pkill: no process found matching '{}'",
                args.first().map(String::as_str).unwrap_or("")
            ));
        }
        let mut results = Vec::new();
        for pid in pids {
            let r = self.kill_pid(pid, "TERM");
            if !r.stdout.is_empty() {
                results.push(r.stdout.trim().to_string());
            }
        }
        Output::ok(results.join("\n"))
    }

    pub(crate) fn cmd_df(&mut self, args: &[String]) -> Output {
        if let Some(id) = self.attached.clone() {
            if let Some(table) = self.with_guest_mut(&id, |g| crate::virt::df_table(g)) {
                return Output::ok(table);
            }
        }
        let used = size_of(&self.vfs) as u64;
        let total = self.ctx.disk_total.unwrap_or(512_000);
        let avail = total.saturating_sub(used);
        let pct = ((used as f64 / total as f64) * 100.0).round() as u64;
        let pct = pct.min(99);
        let human = args.iter().any(|a| a == "-h" || a == "-H");
        let fmt = |n: u64| -> String {
            if !human {
                format!("{}K", (n as f64 / 1024.0).round().max(0.0) as u64)
            } else if n < 1024 {
                format!("{n}B")
            } else if n < 1024 * 1024 {
                format!("{}K", (n as f64 / 1024.0).round() as u64)
            } else {
                format!("{:.1}M", n as f64 / (1024.0 * 1024.0))
            }
        };
        if human {
            Output::ok(format!(
                "Filesystem      Size  Used Avail Use% Mounted on\n\
                 /dev/sda1      {:>5} {:>5} {:>5}  {:>3}% /",
                fmt(total),
                fmt(used),
                fmt(avail),
                pct
            ))
        } else {
            Output::ok(format!(
                "Filesystem     1K-blocks    Used Available Use% Mounted on\n\
                 /dev/sda1      {:>9} {:>7} {:>9}  {:>3}% /",
                fmt(total),
                fmt(used),
                fmt(avail),
                pct
            ))
        }
    }

    pub(crate) fn cmd_du(&mut self, args: &[String]) -> Output {
        let human = args.iter().any(|a| a == "-h" || a == "-sh" || a == "-hs");
        let summarize = args.iter().any(|a| a == "-s" || a == "-sh" || a == "-hs");
        let paths: Vec<&str> = args
            .iter()
            .map(String::as_str)
            .filter(|a| !a.starts_with('-'))
            .collect();
        let default = ["."];
        let targets: Vec<&str> = if paths.is_empty() {
            default.to_vec()
        } else {
            paths
        };
        let format = |n: usize| -> String {
            if !human {
                (n as f64 / 1024.0).round().max(1.0).to_string()
            } else if n < 1024 {
                format!("{n}B")
            } else if n < 1024 * 1024 {
                format!("{}K", (n as f64 / 1024.0).round() as u64)
            } else {
                format!("{:.1}M", n as f64 / (1024.0 * 1024.0))
            }
        };
        let mut lines = Vec::new();
        for target in &targets {
            let Some(res) = vfs::resolve(&self.vfs, &self.cwd, target, &self.home) else {
                return Output::err(format!(
                    "du: cannot access '{target}': No such file or directory"
                ));
            };
            if summarize || res.node.is_file() || targets.len() > 1 {
                lines.push(format!("{}\t{target}", format(size_of(res.node))));
                continue;
            }
            if let Some(children) = res.node.children() {
                for (name, child) in children {
                    lines.push(format!("{}\t{name}", format(size_of(child))));
                }
            }
            lines.push(format!("{}\t.", format(size_of(res.node))));
        }
        Output::ok(lines.join("\n"))
    }

    pub(crate) fn cmd_netstat(&mut self) -> Output {
        let owned = self.ctx.connections.clone();
        let fallback = default_connections();
        let conns = owned.as_deref().unwrap_or(&fallback);
        let mut lines = vec![
            "Proto Local Address           Foreign Address         State       PID/Program name"
                .to_string(),
        ];
        for c in conns {
            lines.push(format!(
                "{:<5} {:<22} {:<23} {:<11} {}",
                c.proto, c.local, c.remote, c.state, c.proc
            ));
        }
        self.ctx.used_netstat = true;
        Output::ok(lines.join("\n"))
    }

    pub(crate) fn cmd_last(&mut self) -> Output {
        self.ctx.used_last = true;
        let base = "root     tty1         Fri Aug 14 19:02   still logged in\n\
                    chief    pts/0        Fri Aug 14 18:55   still logged in\n\
                    miller   pts/2        Fri Aug 14 17:10 - 17:44  (00:34)\n";
        Output::ok(format!("{base}{}", self.ctx.last_log))
    }

    pub(crate) fn cmd_who(&mut self) -> Output {
        Output::ok(
            "root     tty1         2026-08-14 19:02\nchief    pts/0        2026-08-14 18:55",
        )
    }

    pub(crate) fn cmd_crontab(&mut self, args: &[String]) -> Output {
        if !args.iter().any(|a| a == "-l") && args.first().map(String::as_str) != Some("-l") {
            return Output::err("crontab: usage: crontab -l");
        }
        let mut out = String::new();
        if let Some(file) = vfs::resolve(&self.vfs, "/", "/etc/crontab", &self.home) {
            if let Some(c) = file.node.content() {
                out.push_str(c);
            }
        }
        if let Some(dir) = vfs::resolve(&self.vfs, "/", "/etc/cron.d", &self.home) {
            if let Some(children) = dir.node.children() {
                for (name, node) in children {
                    out.push_str(&format!(
                        "\n# /etc/cron.d/{name}\n{}",
                        node.content().unwrap_or("")
                    ));
                }
            }
        }
        self.ctx.used_crontab = true;
        Output::ok(out)
    }

    pub(crate) fn cmd_passwd(&mut self, args: &[String]) -> Output {
        let target = args.first().map(String::as_str).unwrap_or("root");
        if let Some(allow) = self.ctx.allow_passwd_for.clone() {
            if target == allow || (args.is_empty() && allow == "root") {
                self.ctx.password_changed.push(allow.clone());
                return Output::ok(format!(
                    "Changing password for {allow}.\nNew password: ********\npasswd: password updated successfully"
                ));
            }
        }
        if target == "coffee" || target == "beantek" {
            self.ctx.password_changed.push("coffee".into());
            return Output::ok(
                "Changing password for coffee.\nNew password: ********\npasswd: password updated successfully",
            );
        }
        Output::err(
            "passwd: Authentication token manipulation error\n(This console only lets you rotate the appliance account.)",
        )
    }

    pub(crate) fn cmd_ping(&mut self, args: &[String]) -> Output {
        let raw = args.first().map(|s| s.trim()).unwrap_or("");
        if raw.is_empty() || raw.starts_with('-') {
            return Output::err("usage: ping HOST");
        }
        let Some(rec) = lan::catalog(raw) else {
            return Output::err(format!("ping: {raw}: Name or service not known"));
        };
        Output::ok(format!(
            "PING {id} ({addr}) 56(84) bytes of data.\n\
             64 bytes from {id} ({addr}): icmp_seq=1 ttl=64 time=0.4 ms\n\
             64 bytes from {id} ({addr}): icmp_seq=2 ttl=64 time=0.3 ms\n\
             --- {id} ping statistics ---\n\
             2 packets transmitted, 2 received, 0% packet loss",
            id = rec.id,
            addr = rec.addr
        ))
    }
}
